src/source.ts

```diff
@@ -27,68 +28,99 @@ export default class Source extends React.Component<Props, void> {
 
   private id = this.props.id;
 
-  public componentWillMount() {
+  private onSourceData = (e: any) => {
     const { map } = this.context;
-    const { geoJsonSource, tileJsonSource, onSourceAdded } = this.props;
-
-    if (!map.getSource(this.id) && (geoJsonSource || tileJsonSource)) {
-      map.addSource(this.id, (geoJsonSource || tileJsonSource) as any);
-      map.on('load', this.onData);
-
-      if (onSourceAdded) {
-        onSourceAdded(map.getSource(this.id) as GeoJSONSource | TilesJson);
+    const { id, props } = this;
+    if (e.isSourceLoaded && e.sourceId === this.props.id) {
+      // Refresh data (sometimes unloade without apparent reason)
+      const source = map.getSource(id) as GeoJSONSource | TilesJson;
+      if (source && props.geoJsonSource && props.geoJsonSource.data) {
+        const geoJSONSource = source as GeoJSONSource;
+        geoJSONSource.setData(props.geoJsonSource.data);
       }
+      if (source && props.onSourceLoaded) {
+        props.onSourceLoaded(source);
+      }
+      map.off('sourcedata', this.onSourceData);
     }
   }
 
-  private onData = (event: any) => {
+  private onMapLoaded = () => {
+    const { id, props } = this;
     const { map } = this.context;
-
-    const source = map.getSource(this.props.id) as GeoJSONSource;
-
-    const { onSourceLoaded } = this.props;
-    if (source && onSourceLoaded) {
-      onSourceLoaded(source);
-    }
-    // Will fix datasource being empty
-    if (source && this.props.geoJsonSource && this.props.geoJsonSource.data) {
-      source.setData(this.props.geoJsonSource.data as SourceOptionData);
+    const { geoJsonSource, tileJsonSource, onSourceAdded } = props;
+    if (tileJsonSource) {
+      const source = map.getSource(id) as TilesJson;
+      if (!source) {
+        map.addSource(id, tileJsonSource as any);
+        map.on('sourcedata', this.onSourceData);
+        if (onSourceAdded) {
+          onSourceAdded(map.getSource(id) as GeoJSONSource | TilesJson);
+        }
+      }
     }
-    map.off('load', this.onData);
-  }
 
-  public componentWillUnmount() {
-    const { map } = this.context;
-    if (map.getSource(this.id)) {
-      map.removeSource(this.id);
+    // Update geoJsonSource data
+    if (geoJsonSource) {
+      const source = map.getSource(id) as GeoJSONSource;
+      if (!source) {
+        map.addSource(id, geoJsonSource as GeoJSONSource);
+        map.on('sourcedata', this.onSourceData);
+        if (onSourceAdded) {
+          onSourceAdded(map.getSource(id) as GeoJSONSource | TilesJson);
+        }
+      }
     }
   }
 
-  public componentWillReceiveProps(props: Props) {
-    const { id } = this;
+  public shouldComponentUpdate(nextProps: any) {
```

Comment:
I think shouldComponentUpdate is not necessary, I am trying to move away from doing deep equality check on `shouldComponentUpdate` because most of the time we don't need it